pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -61,172 +103,188 @@ type ecrTokenGetter struct {
+		Credentials: nil,
+		Region:      aws.String(region),
+	}))}
+	getter.svc.Handlers.Build.PushFrontNamed(request.NamedHandler{
+		Name: "k8s/user-agent",
+		Fn:   request.MakeAddToUserAgentHandler("kubernetes", version.Get().String()),
+	})
+	getter.svc.Handlers.Sign.PushFrontNamed(request.NamedHandler{
+		Name: "k8s/logger",
+		Fn:   awsHandlerLogger,
+	})
+	return getter
 }
 
-var _ credentialprovider.DockerConfigProvider = &lazyEcrProvider{}
+/*CACHE*/
+// type TimeFunc func(obj interface{}) (time.Time, error)
 
-// ecrProvider is a DockerConfigProvider that gets and refreshes 12-hour tokens
-// from AWS to access ECR.
-type ecrProvider struct {
-	region    string
-	regionURL string
-	getter    tokenGetter
+// type expirationPolicy interface {
+
+type cacheEntry struct {
+	expiresAt   time.Time
+	credentials credentialprovider.DockerConfigEntry
+	host        string
 }
 
-var _ credentialprovider.DockerConfigProvider = &ecrProvider{}
+// ecrExpirationPolicy implements ExpirationPolicy.
+type ecrExpirationPolicy struct{}
 
-// registryURL has different suffix in AWS China region
-func registryURL(region string) string {
-	dnsSuffix := awsStandardDNSSuffix
-	// deal with aws none standard regions
-	if strings.HasPrefix(region, awsChinaRegionPrefix) {
-		dnsSuffix = awsChinaDNSSuffix
-	}
-	return fmt.Sprintf(registryURLTemplate, region, dnsSuffix)
-}
+// func newExpirationPolicy() *ecrExpirationPolicy {
+// 	return &ecrExpirationPolicy{
+// 		expiresAtFunc: expiresAtFunc,
+// 	}
+// }
 
-// RegisterCredentialsProvider registers a credential provider for the specified region.
-// It creates a lazy provider for each AWS region, in order to support
-// cross-region ECR access. They have to be lazy because it's unlikely, but not
-// impossible, that we'll use more than one.
-// This should be called only if using the AWS cloud provider.
-// This way, we avoid timeouts waiting for a non-existent provider.
-func RegisterCredentialsProvider(region string) {
-	klog.V(4).Infof("registering credentials provider for AWS region %q", region)
+// func expiresAtFunc(obj interface{}) (time.Time, error) {
+// 	expiresAt := obj.(cacheEntry).expiresAt
+// 	return expiresAt, nil
+// }
 
-	credentialprovider.RegisterCredentialProvider("aws-ecr-"+region,
-		&lazyEcrProvider{
-			region:    region,
-			regionURL: registryURL(region),
-		})
+// stringKeyFunc is a string as cache key function
+func stringKeyFunc(obj interface{}) (string, error) {
+	key := obj.(cacheEntry).host
+	return key, nil
 }
 
-// Enabled implements DockerConfigProvider.Enabled for the lazy provider.
-// Since we perform no checks/work of our own and actualProvider is only created
-// later at image pulling time (if ever), always return true.
-func (p *lazyEcrProvider) Enabled() bool {
-	return true
+func (p *ecrExpirationPolicy) IsExpired(entry *cache.TimestampedEntry) bool {
+	expiresAt := entry.Obj.(cacheEntry).expiresAt
+	return expiresAt.Before(time.Now()) //TODO clear before expiration
```

Comment:
When should we have it clear? Or should it be at expiresAt?

Reply:
@nckturner this was the refresh discussion. One thing is that I think it would have to be a duration between the time going in and expiresAt using time.Sub as the TTL since I can't think of how else to get it accurately.

Reply:
Per our discussion offline you will definitely want to refresh these credentials before the expiration time, as in do something like: 
```
expiresAt.Add(-time.Hour).Before(time.Now())
```